Frontend/src/Components/Graph/CustomGraph.js

```diff
@@ -1,427 +1,496 @@
+ * @param {number} props.secondsRetained the number of seconds to retain a point on this grpah
+ * @returns
+ */
+export default function CustomGraph(props) {
+  // destructure props
+  const {
+    onSave,
+    title,
+    categories,
+    packedData,
+    initialDatasets,
+    allDatasets,
+    secondsRetained,
+    latestTime,
+    ...stackProps
+  } = props;
+
+  // useEffect(() => {
+  //   console.log(
+  //     packedData
+  //       .find((data) => data.key === "speed")
+  //       .data.map((point) => point.x)
+  //       .join(", ")
+  //   );
+  // }, [packedData]);
+
+  // disclosure for the data selection modal
+  const {
+    isOpen: isDataSelectOpen,
+    onOpen: onDataSelectOpen,
+    onClose: onDataSelectClose,
+  } = useDisclosure();
+  // disclosure for the name modal
+  const {
+    isOpen: isNameModalOpen,
+    onOpen: onNameModalOpen,
+    onClose: onNameModalClose,
+  } = useDisclosure();
+
+  // an object that contains every selected dataset and a boolean to store whether it is enabled
+  const [datasets, setDatasets] = useState(() => {
+    const output = {};
+    for (const dataset of initialDatasets) {
+      output[dataset] = true;
+    }
+
+    // console.log("init datasets:", output, "from", initialDatasets);
+
+    return output;
+  });
+  // useEffect(() => {
+  //   console.log(`datasets for "${title}" updated:`, datasets);
+  // }, [datasets, title]);
+
+  // the data to pass to the graph
+  const [formattedDatasets, setFormattedDatasets] = useState([]);
+  // the keys of the dataset array
+  const [datasetKeys, setDatasetKeys] = useState(Object.keys(datasets));
+  // called when the name is to be saved, memoized for performance
+  const onNameSave = useCallback(
+    (name, isNew) => onSave(name, isNew, datasetKeys),
+    [onSave, datasetKeys]
+  );
+  const [historyLength, setHistoryLength] = useState(
+    secondsRetained ?? GraphData.defaultHistoryLength
+  );
+
+  // side-effects for when new datasets are chosen
+  useEffect(() => {
+    // console.log("yo, new datasets dropped:", datasetKeys);
+    // console.log("keys:", keys);
+
+    // update dataset object {key: boolean}
+    const newDatasets = {};
+    for (const key of datasetKeys) {
+      newDatasets[key] = datasets[key] ?? true;
+    }
+    // console.log("new datasets:", newDatasets);
+    setDatasets(newDatasets);
+  }, [datasetKeys]);
+
+  useEffect(() => {
+    // update formatted datasets storage array if all datasets are updated
+    const formattedDatasets = [];
+    for (const key of datasetKeys) {
+      const temp = packedData.find(
+        (packedDataset) => packedDataset.key === key
+      );
+      if (temp) {
+        const copy = Object.assign({}, temp);
+        copy.hidden = !datasets[key];
+        // console.log(`${title} pushed:`, copy);
+        formattedDatasets.push(copy);
+      } else {
+        console.warn(`unable to find "${key}" for "${title}"`);
+      }
```

Comment:
Is this here to avoid referencing the same dataset in different graphs? Either way, there should be a comment to explain the purpose of copying temp before pushing it to formattedDatasets

Reply:
It is actually to maintain function purity. Good catch, though, so I added a comment